File: sync-service/lib/crdts/DotKernel.ts
import { DotContext } from "./DotContext";

// create a generic tuple type of type T, K, number

// (Value, ID, Dot) -> P(E × I × N)
type dot<T> = [T, number];

class DotKernel<K, T> {
    entryMap: Map<dot<T>, K> = new Map();
    ctx = new DotContext<T>();

    constructor(map?: Map<dot<T>, K>) {
        this.entryMap = map ?? new Map();
    }

    add(id: T, value: K) {
        const dot: dot<T> = this.ctx.makeDot(id);
        // const dot = this.ctx.next(id);
        this.entryMap.set(dot, value)
    }

    remove(value: K) : void {
        for (const [dot, currValue] of this.entryMap) {
            console.log(this.entryMap);
            if (currValue === value) {
                this.removeDot(dot);
            }
        }
    }

    removeDot(dot: dot<T>) : void {
        const [id, dotVal] = dot;
        for (const [arr, ] of this.entryMap) {
            const [currID, currDot] = arr;
            if (currID === id && currDot === dotVal) {
                this.entryMap.delete(arr);
                console.log(this.entryMap);
            }
        }
    }

    has(dot?: dot<T>, value?: K): boolean {
        if (dot) {
            if (value) {
                return this.entryMap.get(dot) === value;
            }
            return this.entryMap.has(dot);
        } else if (value) {
            for (const [, currValue] of this.entryMap) {
                if (currValue === value) {
                    return true;
                }
            }
        }
        return false;
    }


    reset() {
        this.entryMap.clear();
    }

    merge(aw: DotKernel<K, T>): void {
        this.ctx.merge(aw.ctx);

        for (const [arr, value] of aw.entryMap) {
            const [id, dot] = arr
            // new values -> the key is not in this set and it is not in the context
            if (!this.ctx.has(id) && !this.has([id, dot], value)) {
                this.ctx.next(id);
                this.add(id, value);
            }

            this.entryMap.set([id, dot], value);
        }

        for (const [arr, value] of this.entryMap) {
            const [id, dot] = arr
            // removed values -> the key is in the other set's context
            // but it is not in the other set
            if (aw.ctx.has(id) && !aw.has([id, dot], value)) {
                this.ctx.next(id);
                this.remove(id, value);
            }
        }
    }

    toJSON() {
        return {
            value: [...this.entryMap.entries()],
            dots: this.ctx,
        };
    }

    toString(): string {
        let output = "DotKernel:";
        output += " ( ";
        for (const [[id, dot], value] of this.entryMap) {
            output += `\n\t${value}:${id}:${dot},`;
        }
        output += ")";

        return output;
    }

    values(): Array<K> {
        return Array.from(this.entryMap).map(([, value]) => value);
    }
}

export { DotKernel };

File: sync-service/lib/crdts/dotkernel.cpp
template <typename T, typename K>
class dotkernel
{
public:

  map<pair<K,int>,T> ds;  // Map of dots to vals

  dotcontext<K> cbase;
  dotcontext<K> & c;

  // if no causal context supplied, used base one
  dotkernel() : c(cbase) {} 
  // if supplied, use a shared causal context
  dotkernel(dotcontext<K> &jointc) : c(jointc) {} 
//  dotkernel(const dotkernel<T,K> &adk) : c(adk.c), ds(adk.ds) {}

  dotkernel<T,K> & operator=(const dotkernel<T,K> & adk)
  {
    if (&adk == this) return *this;
    if (&c != &adk.c) c=adk.c; 
    ds=adk.ds;
    return *this;
  }

  friend ostream &operator<<( ostream &output, const dotkernel<T,K>& o)
  { 
    output << "Kernel: DS ( ";
    for (const auto & dv : o.ds)
      output <<  dv.first.first << ":" << dv.first.second << 
        "->" << dv.second << " ";
    output << ") ";

    cout << o.c;

    return output;            
  }

  void join (const dotkernel<T,K> & o)
  {
    if (this == &o) return; // Join is idempotent, but just dont do it.
    // DS
    // will iterate over the two sorted sets to compute join
    //typename  map<pair<K,int>,T>::iterator it;
    //typename  map<pair<K,int>,T>::const_iterator ito;
    auto it=ds.begin(); auto ito=o.ds.begin();
    do 
    {
      if ( it != ds.end() && ( ito == o.ds.end() || it->first < ito->first))
      {
        // dot only at this
        if (o.c.dotin(it->first)) // other knows dot, must delete here 
          ds.erase(it++);
        else // keep it
          ++it;
      }
      else if ( ito != o.ds.end() && ( it == ds.end() || ito->first < it->first))
      {
        // dot only at other
        if(! c.dotin(ito->first)) // If I dont know, import
          ds.insert(*ito);
        ++ito;
      }
      else if ( it != ds.end() && ito != o.ds.end() )
      {
        // dot in both
        ++it; ++ito;
      }
    } while (it != ds.end() || ito != o.ds.end() );
    // CC
    c.join(o.c);
  }

  void deepjoin (const dotkernel<T,K> & o)
  {
    if (this == &o) return; // Join is idempotent, but just dont do it.
    // DS
    // will iterate over the two sorted sets to compute join
    //typename  map<pair<K,int>,T>::iterator it;
    //typename  map<pair<K,int>,T>::const_iterator ito;
    auto it=ds.begin(); auto ito=o.ds.begin();
    do 
    {
      if ( it != ds.end() && ( ito == o.ds.end() || it->first < ito->first))
      {
        // dot only at this
        if (o.c.dotin(it->first)) // other knows dot, must delete here 
          ds.erase(it++);
        else // keep it
          ++it;
      }
      else if ( ito != o.ds.end() && ( it == ds.end() || ito->first < it->first))
      {
        // dot only at other
        if(! c.dotin(ito->first)) // If I dont know, import
          ds.insert(*ito);
        ++ito;
      }
      else if ( it != ds.end() && ito != o.ds.end() )
      {
        // dot in both
        // check it payloads are diferent 
        if (it->second != ito->second)
        {
          // if payloads are not equal, they must be mergeable
          // use the more general binary join
          it->second=::join(it->second,ito->second);
        }
        ++it; ++ito;
      }
    } while (it != ds.end() || ito != o.ds.end() );
    // CC
    c.join(o.c);
  }


  dotkernel<T,K> add (const K& id, const T& val) 
  {
    dotkernel<T,K> res;
    // get new dot
    pair<K,int> dot=c.makedot(id);
    // add under new dot
    ds.insert(pair<pair<K,int>,T>(dot,val));
    // make delta
    res.ds.insert(pair<pair<K,int>,T>(dot,val));
    res.c.insertdot(dot);
    return res;
  }

  // Add that returns the added dot, instead of kernel delta
  pair<K,int> dotadd (const K& id, const T& val)
  {
    // get new dot
    pair<K,int> dot=c.makedot(id);
    // add under new dot
    ds.insert(pair<pair<K,int>,T>(dot,val));
    return dot;
  }

  dotkernel<T,K> rmv (const T& val)  // remove all dots matching value
  {
    dotkernel<T,K> res;
    //typename  map<pair<K,int>,T>::iterator dsit;
    for(auto dsit=ds.begin(); dsit != ds.end();)
    {
      if (dsit->second == val) // match
      {
        res.c.insertdot(dsit->first,false); // result knows removed dots
        ds.erase(dsit++);
      }
      else
        ++dsit;
    }
    res.c.compact(); // Maybe several dots there, so atempt compactation
    return res;
  }

  dotkernel<T,K> rmv (const pair<K,int>& dot)  // remove a dot 
  {
    dotkernel<T,K> res;
    auto dsit=ds.find(dot);
    if (dsit != ds.end()) // found it
    {
      res.c.insertdot(dsit->first,false); // result knows removed dots
      ds.erase(dsit++);
    }
    res.c.compact(); // Atempt compactation
    return res;
  }

  dotkernel<T,K> rmv ()  // remove all dots 
  {
    dotkernel<T,K> res;
    for (const auto & dv : ds) 
      res.c.insertdot(dv.first,false);
    res.c.compact();
    ds.clear(); // Clear the payload, but remember context
    return res;
  }

};